### PR TITLE
fix(gatsby) api hooks: typescript return types

### LINKED ISSUE #24229: Wrong return types for some gatsby-ssr hooks
## Description

Some hooks like `onRenderBody` expect to return `any`, when the documentation shows an example returning `void` (no return statement).

### Steps to reproduce

- Create a TS file with 
```ts
import type { GatsbySSR } from "gatsby"

import React from "react"

export const onRenderBody: GatsbySSR["onRenderBody"] = ({ setHeadComponents }) => {
  setHeadComponents([<link key="test" />])
}
```

- Import it in `gatsby-ssr.js` like:
```ts
exports.onRenderBody = require("./src/gatsby/onRenderBody").onRenderBody
```

You might need to add `require("ts-node").register()` to `gatsby-config.js` to make TS work.
Or you can just start from https://github.com/antoinerousseau/gatsby-starter-antoine

### Expected result

No error. Like in the doc: https://www.gatsbyjs.org/docs/ssr-apis/#onRenderBody

### Actual result

TS error: `Type 'void' is not assignable to type 'Promise<any>'` (or `to type 'any'`)

### Environment

```
  System:
    OS: macOS 10.15.3
    CPU: (8) x64 Intel(R) Core(TM) i5-8259U CPU @ 2.30GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 13.12.0 - ~/.nvm/versions/node/v13.12.0/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v13.12.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 81.0.4044.138
    Firefox: 75.0
    Safari: 13.0.5
  npmPackages:
    gatsby: ^2.20.23 => 2.21.39
    gatsby-image: ^2.3.3 => 2.4.4
    gatsby-plugin-google-analytics: ^2.2.3 => 2.3.1
    gatsby-plugin-manifest: ^2.3.4 => 2.4.5
    gatsby-plugin-netlify: ^2.2.2 => 2.3.2
    gatsby-plugin-react-helmet: ^3.2.3 => 3.3.1
    gatsby-plugin-sharp: ^2.5.5 => 2.6.5
    gatsby-plugin-styled-components: ^3.2.2 => 3.3.1
    gatsby-plugin-typescript: ^2.3.2 => 2.4.2
    gatsby-source-filesystem: ^2.2.3 => 2.3.4
    gatsby-transformer-sharp: ^2.4.5 => 2.5.2
```

### PR DESCRIPTION
**Pull request by antoinerousseau**
###  Description

Some hooks expect to return `any`, when they are not expected to return anything. Forcing us to return something, even if it's `null`, when implementing them in TypeScript.

### Documentation

https://www.gatsbyjs.org/docs/ssr-apis/

## Related Issues

Fixes #24229.

### CODE REVIEW

File: packages/gatsby/index.d.ts [hunk 1]
```diff
@@ -394,16 +394,16 @@ export interface GatsbyNode {
...
     args: SetFieldsOnGraphQLNodeTypeArgs,
     options: PluginOptions
-  ): Promise<any>
+  ): Promise<GraphQLFieldConfigMap<any, any>>
   setFieldsOnGraphQLNodeType?(
     args: SetFieldsOnGraphQLNodeTypeArgs,
     options: PluginOptions,
     callback: PluginCallback
-  ): void
+  ): GraphQLFieldConfigMap<any, any>
```
Comment (pieh) [inline packages/gatsby/index.d.ts:406]: for `callback` cases - return should be `void` - we instead use results from callback args, so potentially this should be something like this (and `PluginCallback` should be able to handle accepting result types, as it doesn't right now):

```ts
  setFieldsOnGraphQLNodeType?(
    args: SetFieldsOnGraphQLNodeTypeArgs,
    options: PluginOptions,
    callback: PluginCallback<GraphQLFieldConfigMap<any, any>>
  ): void
```

Comment (antoinerousseau) [reply packages/gatsby/index.d.ts]: Right, thanks, I went ahead and made that return type mandatory to improve readability (`callback?: PluginCallback<void>`).

**Comment by antoinerousseau** [comment]
up

**Comment by antoinerousseau** [comment]
😳

**Comment by antoinerousseau** [comment]
@LekoArts @pieh etc: is the team overwhelmed by the number or PRs? Anything I can do about this?

File: packages/gatsby/index.d.ts [hunk 2]
```diff
@@ -11,7 +11,8 @@ import {
   ComposeScalarTypeConfig,
   ComposeUnionTypeConfig,
 } from "graphql-compose"
-import { GraphQLOutputType } from "graphql"
+import { GraphQLOutputType, GraphQLFieldConfigMap } from "graphql"
+import { PluginRef } from "./src/bootstrap/load-plugins/types"
```
Comment (pieh) [inline packages/gatsby/index.d.ts:15]: @antoinerousseau This `PluginRef` type import seems to cause some conflicts ( there is `PluginRef` type defined in this file already)

**Review by pieh — CHANGES_REQUESTED** [review]
> is the team overwhelmed by the number or PRs?

A little bit, yes. I'm sorry about that :(

If you are still interested in getting this PR across the finish line I left 1 comment. Something weird happened to this PR because it now can't find the remote branch:
Screenshot 2020-09-28 at 10 48 10

Because of above I can't push minor change myself here to get the PR to pass CI tests :( not sure if it's fixable from your side - if not we can reopen PR with new branch with cherry-picked commits from this one.

**Comment by antoinerousseau** [comment]
@pieh: thanks for the feedback! I don't know why the remote branch got lost because it was still present really, but anyway, I opened another PR at #27149 if you want to take a look :) 

Cheers

**Pull request closed without merging**